std/file.d

```diff
@@ -465,54 +465,248 @@ version (linux) @safe unittest
+            }
+            case utf16le:
+            {
+                version(BigEndian)
+                    throw new UTFException("BOM is for UTF-16 BE on Little Endian machine");
+                else
+                    break;
+            }
+            case utf32be:
+            case utf32le: throw new UTFException("UTF-8 requested. BOM is for UTF-32");
+            default: break;
+        }
+    }
+    else
+    {
+        with(BOM) switch (bom)
+        {
+            case utf8: throw new UTFException("UTF-16 requested. BOM is for UTF-8");
+            case utf16be:
+            case utf16le: throw new UTFException("UTF-8 requested. BOM is for UTF-16");
+            case utf32be:
+            {
+                version(BigEndian)
+                    break;
+                else
+                    throw new UTFException("BOM is for UTF-32 LE on Big Endian machine");
+            }
+            case utf32le:
+            {
+                version(BigEndian)
+                    throw new UTFException("BOM is for UTF-32 BE on Little Endian machine");
+                else
+                    break;
+            }
+            default: break;
+        }
+    }
+
+    if (data.length % ElementEncodingType!S.sizeof != 0)
+        throw new UTFException(format!"The content of %s is not UTF-%s"(filename, ElementEncodingType!S.sizeof * 8));
+
+    auto result = trustedCast!S(data);
     validate(result);
     return result;
 }
 
 ///
 @safe unittest
 {
-    import std.exception : enforce;
-    write(deleteme, "abc"); // deleteme is the name of a temporary file
-    scope(exit) remove(deleteme);
-    string content = readText(deleteme);
-    enforce(content == "abc");
+    import std.algorithm.searching : skipOver;
+    import std.exception : assertThrown;
+    import std.utf : UTFException;
+    {
+        write(deleteme, "abc"); // deleteme is the name of a temporary file
+        scope(exit) remove(deleteme);
+        string content = readText(deleteme);
+        assert(content == "abc");
+        // Strips BOM if present.
+        content.skipOver('\uFEFF');
+        assert(content == "abc");
+    }
+    {
+        write(deleteme, "abc"w);
+        scope(exit) remove(deleteme);
+        auto content = readText!wstring(deleteme);
+        assert(content == "abc"w);
+        // Strips BOM if present.
+        content.skipOver('\uFEFF');
+        assert(content == "abc"w);
+    }
+    {
+        write(deleteme, "\uFEFFabc"w); // With BOM
+        scope(exit) remove(deleteme);
+        auto content = readText!wstring(deleteme);
+        assert(content == "\uFEFFabc"w);
+        // Strips BOM if present.
+        content.skipOver('\uFEFF');
+        assert(content == "abc"w);
+    }
+    {
+        write(deleteme, "abc");
+        scope(exit) remove(deleteme);
+        // Throws because the file is not valid UTF-16.
+        assertThrown!UTFException(readText!wstring(deleteme));
+    }
```

Comment:
This example is a bit of a big, dense ball of code. Some whitespace and comments describing what is being tested / presented would be nice. I don't think we need two `skipOver` examples, too.

Reply:
Does this look better then?

Reply:
Yes; I'd still prefix each code block with a one-line comment of what's being tested. E.g.:

```
// Read UTF-8 text file:
{ ... }
// Attempt to read invalid UTF-16 file:
{ ... }
// Read UTF-8 file with BOM:
{ ... }
```

I still don't think we need two `skipOver` examples.

Reply:
One shows it with a BOM and one without, which makes it clearer that it does nothing when there is no BOM while still showing that it does something when there is a BOM.

Reply:
Yeah, but I don't think that's enough value per LOC. People really shouldn't need to play "spot the difference" with our examples.

Reply:
I don't agree, but I don't care enough to keep it if you really don't want it there.

Reply:
How about splitting them up into individual unittests? It will look better on the docs and you can use the comment directly on the unittest:

```d
/// Read UTF-8 ...
unittest
...
```

Reply:
> I don't agree, but I don't care enough to keep it if you really don't want it there.

No, I don't feel strongly about it.

> How about splitting them up into individual unittests?

I think that would be best. I was thinking that duplicating imports for each block would be ugly, but that's no longer the case.

Reply:
I don't really agree that splitting them up is better, but I just did it. So, hopefully this is ready to go now.